edx_when/api.py

```diff
@@ -166,19 +218,29 @@ def get_dates_for_course(course_id, user=None, use_cached=True, schedule=None):
     # to cache invalidation in clear_dates_for_course. This is only safe to do
     # because a) we serialize to cache with pickle; b) we don't write to
     # ContentDate in this function; This is not a great long-term solution.
-    external_cache_key = _content_dates_cache_key(course_id, rel_lookup)
-    qset = cache.get(external_cache_key) if use_cached else None
+    raw_results_cache_key = _content_dates_cache_key(
+        course_id, rel_lookup, subsection_and_higher_only, published_version
+    )
+    qset = None
+    if use_cached:
+        cached_response = TieredCache.get_cached_response(raw_results_cache_key)
+        if cached_response.is_found:
+            qset = cached_response.value
     if qset is None:
+        qset = models.ContentDate.objects.filter(course_id=course_id, active=True, **rel_lookup)
+        if subsection_and_higher_only:
+            # Include NULL block_type values as well because of lazy rollout.
+            qset = qset.filter(Q(block_type__in=('course', 'chapter', 'sequential', None)))
```

Comment:
Do we have a condition on which we can remove the None? Or just always keep because we won't ever be able to guarantee everything has it filled?

Reply:
This PR adds the new column with `null=True` to avoid DB locking on deployment. And I've chosen to do a lazy load of the column upon course publish. With this strategy, it's possible that some courses will never be published and won't have values for block_type ever - unless we populate the value for all courses using some asynchronous process. Once a course *is* published, no `None` values would be expected for block_type.

Summary: With the current plan, just always keep.